### PR TITLE
Implementing discussed 0.2.0 CSI spec changes

### PR DESCRIPTION
**Pull request by sbezverk**
Closes: #172 #173 #174 

Don't use unsigned integers in proto
Change “ID” to “Id” in Method names
Every “noun” should have a message, so Volume should be a message
These are breaking changes from the 0.1 release.

### CODE REVIEW

**Comment by sbezverk** [comment]
@saad-ali @jdef @jieyu  Appologies for bugging you, previous PR stopped accepting any chages I tried to push and I had to open a new one. This PR addresses the same from PR #176

**Comment by jieyu** [comment]
@sbezverk can you add some text saying that negative value MUST not be returned by the SP for each int type?

**Comment by sbezverk** [comment]
@jieyu sure, do you want me to add this comment to each occurence of ```int```  or one common comment would be sufficient?

**Comment by akutz** [comment]
Hi @sbezverk,

Would you mind please titling the PR with a short, succinct description instead of just the issue numbers? Thanks!

**Comment by sbezverk** [comment]
@akutz There are 4 lines summarizing all the changes done by this PR. Do you need more details?

**Comment by sbezverk** [comment]
@akutz oops, misread your comment. Changing title.

**Comment by jieyu** [comment]
@sbezverk I'd prefer each occurence of int to be more explicit.

**Comment by sbezverk** [comment]
@jieyu done, please let me know if added comments are sufficient.

File: csi.proto [hunk 1]
```diff
@@ -70,9 +70,11 @@ message GetSupportedVersionsResponse {
 // Specifies a version in Semantic Version 2.0 format.
 // (http://semver.org/spec/v2.0.0.html)
 message Version {
-  uint32 major = 1;  // This field is REQUIRED.
-  uint32 minor = 2;  // This field is REQUIRED.
-  uint32 patch = 3;  // This field is REQUIRED.
+  // NOTE: Plugin must ensure in returning valid **int** values,
+  // **negative** values must **not** be returned.
```
Comment (saad-ali) [inline csi.proto]: How about: `The value of these fields must not be negative.` for this one.

File: csi.proto [hunk 2]
```diff
@@ -228,19 +230,25 @@ message VolumeCapability {
 message CapacityRange {
   // Volume must be at least this big. This field is OPTIONAL.
   // A value of 0 is equal to an unspecified field value.
-  uint64 required_bytes = 1;
+  // NOTE: Plugin must ensure in returning valid **int** values,
+  // **negative** values must **not** be returned.
```
Comment (saad-ali) [inline csi.proto]: How about: `The value of this field must not be negative.` for this one. And same below?

File: csi.proto (hunk 1 above)
Comment (sbezverk) [reply csi.proto]: what you suggest is to replace second line in the NOTE, right?

File: csi.proto (hunk 2 above)
Comment (sbezverk) [reply csi.proto]: same question as above..

File: csi.proto (hunk 1 above)
Comment (saad-ali) [reply csi.proto]: Correct, I would suggest replace both lines with this.

Comment (sbezverk) [reply csi.proto]: done

File: csi.proto [hunk 3]
```diff
@@ -228,19 +229,22 @@ message VolumeCapability {
 message CapacityRange {
   // Volume must be at least this big. This field is OPTIONAL.
   // A value of 0 is equal to an unspecified field value.
-  uint64 required_bytes = 1;
+  // The value of this field must not be negative.
```
Comment (jieyu) [inline csi.proto]: s/must not/MUST NOT/

File: csi.proto [hunk 4]
```diff
@@ -228,19 +229,22 @@ message VolumeCapability {
...
   // Volume must be at least this big. This field is OPTIONAL.
   // A value of 0 is equal to an unspecified field value.
-  uint64 required_bytes = 1;
+  // The value of this field must not be negative. 
+  int64 required_bytes = 1;
 
   // Volume must not be bigger than this. This field is OPTIONAL.
   // A value of 0 is equal to an unspecified field value.
-  uint64 limit_bytes = 2;
+  // The value of this field must not be negative.
```
Comment (jieyu) [inline csi.proto]: Ditto.

File: csi.proto [hunk 5]
```diff
@@ -228,19 +229,22 @@ message VolumeCapability {
...
 }
 
 // The information about a provisioned volume.
-message VolumeInfo {
+message Volume {
   // The capacity of the volume in bytes. This field is OPTIONAL. If not
   // set (value of 0), it indicates that the capacity of the volume is
   // unknown (e.g., NFS share).
-  uint64 capacity_bytes = 1;
+  // The value of this field must not be negative.
```
Comment (jieyu) [inline csi.proto]: ditto

File: csi.proto [hunk 6]
```diff
@@ -402,7 +406,8 @@ message ListVolumesRequest {
   // in the subsequent `ListVolumes` call. This field is OPTIONAL. If
   // not specified (zero value), it means there is no restriction on the
   // number of entries that can be returned.
-  uint32 max_entries = 2;
+  // The value of this field must not be negative.
```
Comment (jieyu) [inline csi.proto]: Ditto.

File: csi.proto [hunk 7]
```diff
@@ -452,7 +457,8 @@ message GetCapacityResponse {
   // specified in the request, the Plugin SHALL take those into
   // consideration when calculating the available capacity of the
   // storage. This field is REQUIRED.
-  uint64 available_capacity = 1;
+  // The value of this field must not be negative.
```
Comment (jieyu) [inline csi.proto]: Ditto.

File: spec.md [hunk 8]
```diff
@@ -386,9 +386,10 @@ message GetSupportedVersionsResponse {
 // Specifies a version in Semantic Version 2.0 format.
 // (http://semver.org/spec/v2.0.0.html)
 message Version {
-  uint32 major = 1;  // This field is REQUIRED.
-  uint32 minor = 2;  // This field is REQUIRED.
-  uint32 patch = 3;  // This field is REQUIRED.
+  // The value of these fields must not be negative.
```
Comment (jieyu) [inline spec.md]: I'd suggest have one such line per field (in case we add more fields in the future)

**Comment by akutz** [comment]
I think this PR should either contain multiple commits or be split into multiple PRs. I don’t like how many changes are being performed in a single commit. There are clear issues, and this PR doesn’t directly relate to any parictlaur one, which makes tracking things later more difficult.

File: csi.proto (hunk 3 above)
Comment (sbezverk) [reply csi.proto]: done

File: csi.proto (hunk 4 above)
Comment (sbezverk) [reply csi.proto]: done

File: csi.proto (hunk 5 above)
Comment (sbezverk) [reply csi.proto]: done

File: csi.proto (hunk 6 above)
Comment (sbezverk) [reply csi.proto]: done

File: csi.proto (hunk 7 above)
Comment (sbezverk) [reply csi.proto]: done

File: spec.md (hunk 8 above)
Comment (sbezverk) [reply spec.md]: done

**Comment by sbezverk** [comment]
@akutz This change is fairly small it is just 27 lines in spec.md, does it really requires separate commits?

**Comment by akutz** [comment]
I don’t consider 27 lines that small, and separate commits isn’t related to size but the purpose of the change. If changes are not related, why make them part of the same commit? Commits are cheap, and it’s good practice to make reverting changes easier by isolating work into smaller chunks. For example, if this PR is reverted, would you want it pulled out in its entirety or does it make sense to pull out smaller units of work?

-- 
-a

File: csi.proto [hunk 9]
```diff
@@ -228,19 +231,22 @@ message VolumeCapability {
 message CapacityRange {
   // Volume must be at least this big. This field is OPTIONAL.
   // A value of 0 is equal to an unspecified field value.
-  uint64 required_bytes = 1;
+  // The value of this field MUST NOT be negative. 
+  int64 required_bytes = 1;
```
Comment (j-griffith) [inline csi.proto]: Do we want to get rid of the use of underscores while we're at it?  Granted that's mostly my own pedantic annoyance :)  and it would be a separate patch but just asking.

**Comment by sbezverk** [comment]
@j-griffith csi.proto is autogenerated file, the actual change was applied ONLY to spec.md

**Comment by akutz** [comment]
I'm confused as to why we wouldn't used unsigned integers in the protobuf. It makes sense for versions.

File: csi.proto (hunk 9 above)
Comment (akutz) [reply csi.proto]: Hi @j-griffith,

Please leave the underscores. The tell language generators how to camel case things.

**Comment by akutz** [comment]
I don't mind a lot of these changes, but aside from `GetNodeID` -> `NodeGetID`, many of them seem to be simply based on preference, and that's not a good reason IMO to break the existing spec. The `VolumeInfo` to `Volume` one, for example, will have an especially large impact.

**Comment by akutz** [comment]
I answered my own question about the unsigned integers after reading the original issue:

> Recommendation from a gRPC author after review of CSI API: We should not use unsigned integers, and should change them to signed integers because not all languages allow unsigned integers (e.g. Java), and this can cause issues.

**Comment by sbezverk** [comment]
@akutz Based on my understanding, since version info gets send/recevied in gRPC messages, the recommendation from gRPC team was to use int instead of uint as not all gRPC clients support uint type.

**Comment by akutz** [comment]
I still remain adamant that there should be a separate commit, perhaps even PR, for each specific issue. Commits aren't about adding to a project, but about making it easy to remove things later if necessary. As it is now, any issue with any of these changes will require the removal of all of these changes.

**Comment by saad-ali** [comment]
LGTM

I agree it would be nice to have had this PR broken in to multiple commits. But if that's too much of a pain to do at this point, I'm fine accepting it as is.

> I don't mind a lot of these changes, but aside from GetNodeID -> NodeGetID, many of them seem to be simply based on preference, and that's not a good reason IMO to break the existing spec. The VolumeInfo to Volume one, for example, will have an especially large impact.

If we're going to make breaking changes, might as well tighten the spec up. The reasoning behind VolumeInfo vs Volume is to align with the operation names CreateVolume, PublishVolume, etc.

**Comment by sbezverk** [comment]
@jdef appreciate if you could review this revision of the spec. thanks a lot.

**Comment by saad-ali** [comment]
Merging. Please ensure future PRs are broken in to logical commits.

**Comment by akutz** [comment]
Just logging my objection to this being accepted as a single commit. It’s bad form, bad precedent, and bad practice. Reverting any of these now means reverting them all.

**Comment by sbezverk** [comment]
@akutz all 3 changes are a part of 0.2.0, so it is ALL or nothing. You cannot have version 0.1.0.3/4. That is the only reason why these three changes are bundled.

**Comment by akutz** [comment]
That is a blunt approach to a process usually handled with more finesse and care.

All commits after the previous tag are not generally squashed upon a release, just because a release can be considered a sum of its parts. That’s a bad process. Features and changes are atomic so they can be reverted later, if necessary, even after a release. 

This PR and its changes should have been multiple commits to ensure that changes corresponding to each issue could be reverted separately.

This PR was merged against all best practices, hence my objection.
